### PR TITLE
Rewrite OTA update code

### PR DESCRIPTION
**Pull request by basvs**
Major code rewrite / cleanup

### CODE REVIEW

File: components/sha2017/sha2017_ota.c [hunk 1]
```diff
@@ -72,417 +63,503 @@ static const char *REQUEST = "GET " BADGE_OTA_WEB_PATH " HTTP/1.0\r\n"
...
+
+	/* determine partitions */
+	const esp_partition_t *part_running = esp_ota_get_running_partition();
+	ESP_LOGI(TAG, "Running from partition type %d subtype %d (offset 0x%08x)",
+			part_running->type, part_running->subtype, part_running->address);
+
+	const esp_partition_t *part_update = esp_ota_get_next_update_partition(NULL);
+	ESP_LOGI(TAG, "Writing to partition type %d subtype %d (offset 0x%08x)",
+			part_update->type, part_update->subtype, part_update->address);
+	assert(part_update != NULL);
```
Comment (raboof) [inline components/sha2017/sha2017_ota.c]: codacy is right: doing `assert(part_update != NULL);` *after* `part_update->type` on the previous line doesn't make sense :)

File: components/sha2017/sha2017_ota.c [hunk 2]
```diff
@@ -72,417 +63,503 @@ static const char *REQUEST = "GET " BADGE_OTA_WEB_PATH " HTTP/1.0\r\n"
...
+		task_fatal_error();
+	}
+	ESP_LOGI(TAG, "%d bytes written", strlen(REQUEST));
+
+	show_percentage("Starting OTA update", 0, false);
+
+	/* read until we have received the status line */
+	ESP_LOGI(TAG, "Reading HTTP response status line.");
+	uint8_t *crlf;
+	while ((crlf = index_crlf(buffer, buffer_len)) == NULL) {
```
Comment (raboof) [inline components/sha2017/sha2017_ota.c]: Hmm, not sure what codacy is on about here: when after the last iteration `crlf != NULL` then `*crlf = 0;` later should be fine. Not super easy to read though, might warrant a bit of refactoring anyway?

File: components/sha2017/sha2017_ota.c (hunk 1 above)
Comment (basvs) [inline components/sha2017/sha2017_ota.c]: check. fixed in upcoming commit

File: components/sha2017/sha2017_ota.c (hunk 2 above)
Comment (basvs) [inline components/sha2017/sha2017_ota.c]: Could fix this with code duplication (which results in slower code and risking forgetting the 2nd call to index_crlf()) or by splitting this in an ugly 'while (1) { crlf = index(..); if (crlf != NULL) break; ..'

Comment (basvs) [inline components/sha2017/sha2017_ota.c]: I'll add some extra comments and merge the code. If necessary, we can always create a new issue or pull-request for this.

**Pull request merged**